File: js/image_stream.js
import { WebSocketMQClient } from './client.js'

class ImageStream {
  constructor(url = 'ws://localhost:6789', bufferSize = 1) {
    this.url = url
    this.bufferSize = bufferSize
    this.queues = {} // Queues for storing frames for different topics
    this.frames = {} // Current frames for different topics
    this.metadata = {} // Metadata for different topics
    this.subscribersReady = {} // Events for managing subscribers for different topics
    this.decoders = {} // Decoders for different topics
    this.encoders = {} // Encoders for different topics
    this.onReceives = {} // Key: topic, value: callback function (topic, image)
    this.client = new WebSocketMQClient(url)
    this.stopEvent = false
    this.frameCount = {} // Record the frame count for each topic
  }

  async onMessage(topic, payload, props) {
    const contentType = props['content_type'] || ''

    if (contentType === 'application/json') {
      const newMetadata = JSON.parse(payload)
      this.metadata[topic] = newMetadata
      this.subscribersReady[topic] = false
      await this.initializeDecoder(topic)
    } else if (contentType === 'video/encoded') {
      const isKeyframe = !!(new DataView(payload.buffer).getUint8(0))
      if (isKeyframe) {
        this.subscribersReady[topic] = true
      }
      if (this.subscribersReady[topic]) {
        const decoder = this.decoders[topic]
        try {
          const chunk = new EncodedVideoChunk({
            type: isKeyframe ? 'key' : 'delta',
            timestamp: 0,
            data: payload.slice(1)
          })
          const frame = await this.decodeChunk(decoder, chunk)
          this.frames[topic] = frame
          const onReceive = this.onReceives[topic] || null
          if (onReceive) {
            onReceive(topic, frame)
          }
          frame.close() // Ensure the frame is closed after use
        } catch (e) {
          console.error(`Error decoding packet: ${e}`)
        }
      }
    }
  }

  async decodeChunk(decoder, chunk) {
    return new Promise((resolve, reject) => {
      let frame
      decoder.decode(chunk, {
        complete: output => {
          frame = output
          resolve(frame)
        },
        error: reject
      })
    })
  }

  async initializeEncoder(topic, metadata) {
    const encoder = new VideoEncoder({
      output: chunk => this.encodeOutput(chunk, topic),
      error: e => console.error(`Encoder error: ${e}`)
    })
    const init = {
      codec: 'vp09.00.10.08',
      width: metadata.width || 640,
      height: metadata.height || 480,
      bitrate: 5000000,
      framerate: metadata.frame_rate || 30
    }
    await encoder.configure(init)
    this.encoders[topic] = encoder
    this.frameCount[topic] = 0 // Initialize frame count for the topic to 0
  }

  encodeOutput(chunk, topic) {
    const packet = new Uint8Array(chunk.byteLength + 1)
    packet[0] = chunk.type === 'key' ? 1 : 0
    chunk.copyTo(packet.subarray(1))
    this.client.publish(topic, packet, 'video/encoded')
  }

  async initializeDecoder(topic) {
    const decoder = new VideoDecoder({
      output: frame => {
        this.frames[topic] = frame
        const onReceive = this.onReceives[topic] || null
        if (onReceive) {
          onReceive(topic, frame)
        }
      },
      error: e => console.error(`Decoder error: ${e}`)
    })
    const codec = {
      codec: 'vp09.00.10.08'
    }
    await decoder.configure(codec)
    this.decoders[topic] = decoder
  }

  async encodeFrame(image, topic) {
    const metadata = this.metadata[topic]
    if (!metadata) {
      throw new Error(`No metadata available for topic ${topic}`)
    }

    if (!this.encoders[topic]) {
      await this.initializeEncoder(topic, metadata)
    }

    const encoder = this.encoders[topic]

    // Ensure image is an HTMLCanvasElement, HTMLVideoElement, or ImageBitmap
    let frame
    if (image instanceof HTMLCanvasElement || image instanceof HTMLVideoElement || image instanceof ImageBitmap) {
      frame = new VideoFrame(image, { timestamp: Date.now() })
    } else if (image instanceof ImageData) {
      const canvas = document.createElement('canvas')
      canvas.width = image.width
      canvas.height = image.height
      const ctx = canvas.getContext('2d')
      ctx.putImageData(image, 0, 0)
      frame = new VideoFrame(canvas, { timestamp: Date.now() })
    } else {
      throw new Error('Unsupported image type.')
    }

    // Ensure the first frame of each topic is a keyframe and set keyframe according to gop_size
    const gopSize = metadata.gop_size || 50
    const isKeyframe = this.frameCount[topic] % gopSize === 0
    this.frameCount[topic] = (this.frameCount[topic] + 1) % gopSize // Increment frame count and mod by gopSize

    await encoder.encode(frame, { keyFrame: isKeyframe })
    frame.close() // Ensure the frame is closed after encoding
  }

  async sendFrame() {
    while (!this.stopEvent) {
      for (const topic in this.queues) {
        if (this.queues[topic].length > 0) {
          const image = this.queues[topic].shift()
          await this.encodeFrame(image, topic)
        }
      }
      await new Promise(resolve => setTimeout(resolve, 1))
    }
  }

  async start() {
    this.client.connect()
    await new Promise(resolve => setTimeout(resolve, 500))
    this.thread = this.sendFrame()
  }

  async stop() {
    this.stopEvent = true
    await this.thread
    this.client.disconnect()
  }

  subscribe(topic, onReceive = null) {
    if (onReceive) {
      this.onReceives[topic] = onReceive
    }
    this.client.subscribe(topic, (t, p, props) => this.onMessage(t, p, props))
  }

  unsubscribe(topic) {
    delete this.onReceives[topic]
    this.client.unsubscribe(topic)
  }

  publish(topic, metadata) {
    this.metadata[topic] = metadata
    if (this.encoders[topic]) {
      delete this.encoders[topic] // Remove the encoder to ensure it gets re-initialized with new metadata
    }
    this.client.publish(topic, JSON.stringify(metadata), 'application/json')
  }

  addImage(topic, image) {
    if (!this.queues[topic]) {
      this.queues[topic] = []
    }

    if (this.queues[topic].length >= this.bufferSize) {
      this.queues[topic].shift() // Remove the oldest image to maintain buffer size
    }

    this.queues[topic].push(image)
  }

  getImage(topic) {
    return this.frames[topic]
  }

  displayImage(topic, canvas) {
    this.subscribe(topic, (topic, frame) => {
      // Update the canvas size if necessary
      if (canvas.width !== frame.displayWidth || canvas.height !== frame.displayHeight) {
        canvas.width = frame.displayWidth
        canvas.height = frame.displayHeight
      }
      canvas.getContext('2d').drawImage(frame, 0, 0)
      frame.close() // Ensure the frame is closed after use
    })
  }
}

export { ImageStream }

File: python/src/wsmq/image_stream.py
# Note: Only support VP9 now
import av
import threading
import queue
import struct
import time
import io
import json
import numpy as np
from PIL import Image
import cv2
from wsmq import WebSocketMQClient

class ImageStream:
    def __init__(self, url='ws://localhost:6789', buffer_size=1):
        self.url = url
        self.buffer_size = buffer_size
        self.queues = {}  # Queues for storing frames for different topics
        self.frames = {}  # Current frames for different topics
        self.metadata = {}  # Metadata for different topics
        self.stop_event = threading.Event()
        self.client = WebSocketMQClient(url=url)
        self.subscribers_ready = {}  # Events for managing subscribers for different topics
        self.decoders = {}  # Decoders for different topics
        self.encoders = {}  # Encoders for different topics
        self.on_receives = {}  # key: topic, value: (callback function (topic, image), image_format)

    def on_message(self, topic, payload, props):
        content_type = props.get('content_type', '')

        if content_type == 'application/json':  # Metadata
            metadata = json.loads(payload)
            self.metadata[topic] = metadata
            self.subscribers_ready[topic] = threading.Event()
            self.initialize_decoder(topic)
        elif content_type == 'video/encoded':  # Video frame
            is_keyframe = bool(struct.unpack('!B', payload[0:1])[0])
            if is_keyframe:
                self.subscribers_ready[topic].set()
            if self.subscribers_ready.get(topic, threading.Event()).is_set():
                packet = av.packet.Packet(payload[1:])
                if topic not in self.decoders:
                    self.initialize_decoder(topic)
                decoder = self.decoders[topic]
                try:
                    for frame in decoder.decode(packet):
                        self.frames[topic] = frame
                        on_receive, image_format = self.on_receives.get(topic, (None, 'PIL'))
                        if on_receive is not None:
                            image = self.convert_frame(frame, image_format)
                            on_receive(topic, image)
                except av.AVError as e:
                    print(f"Error decoding packet: {e}")

    def convert_frame(self, frame, image_format):
        '''
        Convert frame to specified image format
        image_format = None, 'ndarray', 'opencv', 'PIL'
        '''
        ndarray = frame.to_ndarray(format='rgb24')
        if image_format == 'ndarray':
            return ndarray
        elif image_format == 'opencv':
            return cv2.cvtColor(ndarray, cv2.COLOR_RGB2BGR)
        elif image_format == 'PIL':
            return Image.fromarray(ndarray)
        else:
            raise ValueError(f"Unsupported image format: {image_format}")

    def initialize_encoder(self, topic, metadata):
        '''
        Initialize encoder for the given topic
        '''
        codec_name = 'vp9'
        container = av.open(io.BytesIO(), mode='w', format='webm')
        stream = container.add_stream(codec_name, rate=metadata.get('frame_rate', 30))
        stream.width = metadata.get('width', 640)
        stream.height = metadata.get('height', 480)
        stream.pix_fmt = metadata.get('pix_fmt', 'yuv420p')
        stream.codec_context.gop_size = metadata.get('gop_size', 50)  # Set keyframe interval in frames or called GOP size
        stream.codec_context.options = {'lag-in-frames': '0'} # zero latency
        self.encoders[topic] = stream

    def initialize_decoder(self, topic):
        '''
        Initialize decoder for the given topic
        '''
        if topic in self.decoders:
            del self.decoders[topic]
        codec_name = 'vp9'
        self.decoders[topic] = av.codec.CodecContext.create(codec_name, 'r')

    def encode_frame(self, frame, topic):
        '''
        Encode frame and return packets
        '''
        metadata = self.metadata.get(topic)
        if metadata is None:
            raise ValueError(f"No metadata available for topic {topic}")

        if topic not in self.encoders:
            self.initialize_encoder(topic, metadata)
        stream = self.encoders[topic]

        frame = av.VideoFrame.from_ndarray(frame, format='bgr24')
        packets = stream.encode(frame)
        keyframes = [pkt for pkt in packets if pkt.is_keyframe]
        if keyframes:
            # Let the latter peer get the metadata if encounter key frame
            metadata_json = json.dumps(metadata)
            self.client.publish(topic, metadata_json, content_type='application/json')
        return packets

    def send_frame(self):
        '''
        Thread function to send frames
        '''
        while not self.stop_event.is_set():
            for topic in list(self.queues):
                if topic in self.queues and not self.queues[topic].empty():
                    frame = self.queues[topic].get()
                    packets = self.encode_frame(frame, topic)
                    for packet in packets:
                        self.client.publish(topic, struct.pack('!B', packet.is_keyframe) + bytes(packet), content_type='video/encoded')
            time.sleep(0.001)

    def start(self):
        '''
        Start the client and the frame sending thread
        '''
        self.client.connect()
        time.sleep(0.5)
        self.thread = threading.Thread(target=self.send_frame, daemon=True)
        self.thread.start()

    def stop(self):
        '''
        Stop the frame sending thread and disconnect the client
        '''
        self.stop_event.set()
        self.thread.join()
        self.client.disconnect()
    
    def subscribe(self, topic, on_receive=None, image_format='PIL'):
        '''
        Subscribe to a topic with a callback function and image format
        image_format = None, 'ndarray', 'opencv', 'PIL'
        '''
        if on_receive is not None:
            self.on_receives[topic] = (on_receive, image_format)
        self.client.subscribe(topic, lambda t, p, props: self.on_message(t, p, props))
    
    def unsubscribe(self, topic):
        '''
        Unsubscribe from a topic
        '''
        if topic in self.on_receives:
            del self.on_receives[topic]
        self.client.unsubscribe(topic)

    def publish(self, topic, metadata):
        '''
        Publish metadata for the given topic
        '''
        self.metadata[topic] = metadata
        if topic in self.encoders:
            del self.encoders[topic]
        metadata_json = json.dumps(metadata)
        self.client.publish(topic, metadata_json, content_type='application/json')

    def add_image(self, topic, image, image_format=None):
        '''
        Add image to the queue for the given topic
        image_format = None, 'ndarray', 'opencv', 'PIL'
        '''
        if isinstance(image, Image.Image):
            image = cv2.cvtColor(np.array(image), cv2.COLOR_RGB2BGR)
        elif isinstance(image, np.ndarray):
            if image.ndim == 2:
                image = cv2.cvtColor(image, cv2.COLOR_GRAY2BGR)
            elif image.shape[2] == 3:
                if image_format != 'opencv':
                    image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR)
            elif image.shape[2] == 4:
                if image_format == 'opencv':
                    image = cv2.cvtColor(image, cv2.COLOR_BGRA2BGR)
                elif image_format == 'ndarray':
                    image = cv2.cvtColor(image, cv2.COLOR_RGBA2BGR)
        elif isinstance(image, bytes):
            nparr = np.frombuffer(image, np.uint8)
            image = cv2.imdecode(nparr, cv2.IMREAD_COLOR)

        if topic not in self.queues:
            self.queues[topic] = queue.Queue(maxsize=self.buffer_size)
        if self.queues[topic].full():
            self.queues[topic].get()
        self.queues[topic].put(image)

    def get_image(self, topic, image_format=None):
        '''
        Get the current frame for the given topic
        image_format = None, 'ndarray', 'opencv', 'PIL'
        '''
        frame = self.frames.get(topic)
        if frame:
            if image_format is None: # Original frame
                return frame
            elif image_format == 'ndarray': # NumPy array (RGB)
                return frame.to_ndarray(format='rgb24')
            elif image_format == 'opencv': # OpenCV image (BGR)
                return frame.to_ndarray(format='bgr24')
            elif image_format == 'PIL': # PIL image
                return Image.fromarray(frame.to_ndarray(format='rgb24'))
            else:
                raise ValueError(f"Unsupported image format: {image_format}")
        return None